week-3/CrowdFunding/CrowdFunding.sol

```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.8.17;
+
+contract CrowdFunding {
+
+    // struct to hold contributor details
+    struct Contributor {
+        uint256 donated;
+        address donerId;
+    }
+
+    // struct to hold project details
+    struct Project {
+        uint256 projectId;
+        string projectName;
+        uint256 targetAmount;
+        uint256 collectedAmount;
+        address payable beneficiaryAddress;
+        Contributor[] contributors;
+        bool isOpen;
+    }
+
+    // state variables
+    address admin;
+    uint256 public projectIndex;
+    mapping(uint256 => Project) public projectList;
+
+    // events
+    event contributionAdded(
+        string indexed _projectName,
+        uint256 indexed _donation,
+        address indexed _donor
+    );
+    event fundingTargetAchieved(
+        string indexed _projectName,
+        uint256 indexed _amount,
+        address indexed _beneficiary
+    );
+    event fundsReturned(
+        string indexed _projectName,
+        uint256 indexed _refundAmount,
+        address indexed _refundAddress
+    );
+
+    // set admin address on contract deployment
+    constructor() {
+        admin = msg.sender;
+    }
+
+    // register new project for funding
+    function register(
+        string memory _name,
+        uint256 _target,
+        address payable _beneficiary
+    ) public {
+        Project storage newProject = projectList[projectIndex];
```

Comment:
Gas can be optimized by creating a Project Struct in memory, and updating all the values to it, and then copying the Project struct in memory to the actual storage variable

Reply:
```Project memory newProject = Project(projectIndex, _name, _target, 0, _beneficiary, true);
projectList[projectIndex] = newProject;

projectIndex++;
```